homeassistant/components/mqtt/water_heater.py

```diff
@@ -0,0 +1,438 @@
+            self._attr_target_temperature = config[CONF_TEMP_INITIAL]
+        if self._topic[CONF_MODE_STATE_TOPIC] is None or self._optimistic:
+            self._attr_current_operation = STATE_OFF
+
+        value_templates: dict[str, Template | None] = {}
+        for key in VALUE_TEMPLATE_KEYS:
+            value_templates[key] = None
+        if CONF_VALUE_TEMPLATE in config:
+            value_templates = {
+                key: config.get(CONF_VALUE_TEMPLATE) for key in VALUE_TEMPLATE_KEYS
+            }
+        for key in VALUE_TEMPLATE_KEYS & config.keys():
+            value_templates[key] = config[key]
+        self._value_templates = {
+            key: MqttValueTemplate(
+                template,
+                entity=self,
+            ).async_render_with_possible_json_value
+            for key, template in value_templates.items()
+        }
+
+        self._command_templates = {}
+        for key in COMMAND_TEMPLATE_KEYS:
+            self._command_templates[key] = MqttCommandTemplate(
+                config.get(key), entity=self
+            ).async_render
+
+        support = WaterHeaterEntityFeature(0)
+        if (self._topic[CONF_TEMP_STATE_TOPIC] is not None) or (
+            self._topic[CONF_TEMP_COMMAND_TOPIC] is not None
+        ):
+            support |= WaterHeaterEntityFeature.TARGET_TEMPERATURE
+
+        if (self._topic[CONF_MODE_STATE_TOPIC] is not None) or (
+            self._topic[CONF_MODE_COMMAND_TOPIC] is not None
+        ):
+            support |= WaterHeaterEntityFeature.OPERATION_MODE
+
+        self._attr_supported_features = support
+
+    def _prepare_subscribe_topics(self) -> None:
+        """(Re)Subscribe to topics."""
+        topics: dict[str, dict[str, Any]] = {}
+        qos: int = self._config[CONF_QOS]
+
+        def add_subscription(
+            topics: dict[str, dict[str, Any]],
+            topic: str,
+            msg_callback: Callable[[ReceiveMessage], None],
+        ) -> None:
+            if self._topic[topic] is not None:
+                topics[topic] = {
+                    "topic": self._topic[topic],
+                    "msg_callback": msg_callback,
+                    "qos": qos,
+                    "encoding": self._config[CONF_ENCODING] or None,
+                }
+
+        def render_template(
+            msg: ReceiveMessage, template_name: str
+        ) -> ReceivePayloadType:
+            template = self._value_templates[template_name]
+            return template(msg.payload)
+
+        @callback
+        def handle_temperature_received(
+            msg: ReceiveMessage, template_name: str, attr: str
+        ) -> None:
+            """Handle temperature coming via MQTT."""
+            payload = render_template(msg, template_name)
+
+            if not payload:
+                _LOGGER.debug(
+                    "Invalid empty payload for attribute %s, ignoring update",
+                    attr,
+                )
+                return
```

Comment:
Seems you miss some tests

Reply:
Just to be clear, do you mean missing tests that cover conditions like the one above? If so, consider it done

Reply:
Yeah, you can do a coverage test to see what lines are not covered with tests.

e.g.:

`pytest -n auto tests/components/mqtt/ --cov=homeassistant.components.mqtt --cov-report term-missing`